### PR TITLE
PR: Unify how extra selections are added to the Editor

### PR DESCRIPTION
**Pull request by rlaverde**
Fixes: #5176 

I'm not sure if adding `DecorationManager` to the `TextEditBaseWidget` is ok (commit f408110), or should `DecorationManager` belong to `CodeEditor` and use `try..  except` in `TextEditBaseWidget` methods

### CODE REVIEW

**Comment by goanpeca** [comment]
@rlaverde it depends on where we are using `TextEditBaseWidget` and if in those places we require the functionality. Internal console, history log, I guess? @ccordoba12 ?

**Comment by rlaverde** [comment]
Searching all the codebase, `ExtraSelection` objects were only created in the base `TextEditBaseWidget` class. 

Console and history add extra selections, but they do it through the `TextEditBaseWidget`methods.

**Comment by rlaverde** [comment]
I'm getting tired of restarting the travis python2.7 building, and every time a different test fails (unrelated to this PR)

File: spyder/widgets/panels/codefolding.py [hunk 1]
```diff
@@ -350,7 +350,7 @@ def _get_scope_highlight_color(self):
     def _decorate_block(self, start, end):
         color = self._get_scope_highlight_color()
```
Comment (goanpeca) [inline spyder/widgets/panels/codefolding.py]: A docstring would be nice :-) with some explanation on the args.

File: spyder/widgets/sourcecode/base.py [hunk 2]
```diff
@@ -34,7 +33,8 @@
 from spyder.widgets.calltip import CallTipWidget
 from spyder.widgets.mixins import BaseEditMixin
 from spyder.widgets.sourcecode.terminal import ANSIEscapeCodeHandler
-
+from spyder.widgets.sourcecode.api.decoration import TextDecoration
+from spyder.widgets.sourcecode.utils.decoration import TextDecorationsManager
```
Comment (goanpeca) [inline spyder/widgets/sourcecode/base.py]: Why we have the manager outside of the api, is because we don't need/want to expose that?

File: spyder/widgets/sourcecode/base.py [hunk 3]
```diff
@@ -34,7 +33,8 @@
 from spyder.widgets.calltip import CallTipWidget
 from spyder.widgets.mixins import BaseEditMixin
 from spyder.widgets.sourcecode.terminal import ANSIEscapeCodeHandler
-
+from spyder.widgets.sourcecode.api.decoration import TextDecoration
+from spyder.widgets.sourcecode.utils.decoration import TextDecorationsManager
 
 def insert_text_to(cursor, text, fmt):
     """Helper to print text, taking into account backspaces"""
```
Comment (goanpeca) [inline spyder/widgets/sourcecode/base.py]: period at the end of the docstring ;-)

File: spyder/widgets/sourcecode/base.py [hunk 4]
```diff
@@ -236,7 +236,7 @@ def __init__(self, parent=None):
         BaseEditMixin.__init__(self)
         self.setAttribute(Qt.WA_DeleteOnClose)
         
-        self.extra_selections_dict = OrderedDict()
+        self.extra_selections_dict = {}
```
Comment (goanpeca) [inline spyder/widgets/sourcecode/base.py]: Why the change is not needed? because of the `draworder` parameter?

File: spyder/widgets/sourcecode/base.py [hunk 5]
```diff
@@ -330,30 +332,28 @@ def get_extra_selections(self, key):
         return self.extra_selections_dict.get(key, [])
 
     def set_extra_selections(self, key, extra_selections):
+        # current line has to be highlighted first
```
Comment (goanpeca) [inline spyder/widgets/sourcecode/base.py]: Maybe a docstring would be nice, also stating what key and extra_selections is supposed to be. Like a poors-man typing :-p

File: spyder/widgets/sourcecode/utils/decoration.py [hunk 6]
```diff
@@ -50,6 +49,21 @@ def remove(self, decoration):
         except ValueError:
             return False
 
+    def extend(self, decorations):
```
Comment (goanpeca) [inline spyder/widgets/sourcecode/utils/decoration.py]: I still prefer add :-p

File: spyder/widgets/sourcecode/base.py (hunk 2 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/base.py]: Hmm yes, I don't think there is a reason to expose it

Normally you will create new TextDecoration and append them to the editor.decorations.

File: spyder/widgets/sourcecode/base.py (hunk 3 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/base.py]: I didn't add this docstring 🙈  (I'll improve it with some clarifications of the  new behavior and fix that style error. )

File: spyder/widgets/sourcecode/base.py (hunk 4 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/base.py]: Partially yes, extr selections are now ordered  by the `DecorationManager`,

I moved this logic to `DecorationManager._order_decorations` this will allow that all decorations added to the editor will be order (according draw_order and selection length)

File: spyder/widgets/sourcecode/base.py (hunk 3 above)
Comment (goanpeca) [reply spyder/widgets/sourcecode/base.py]: I know, but its a ood opportunity to clean it up a bit :-)

File: spyder/widgets/sourcecode/base.py (hunk 5 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/base.py]: yep, them I'll improve all docstring of  modified functions (no only the ones that I added)

File: spyder/widgets/sourcecode/utils/decoration.py (hunk 6 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/utils/decoration.py]: `extend` is different to `add`. This function extends a list with another list
Maybe `append` could be changed to `add`

Comment (goanpeca) [reply spyder/widgets/sourcecode/utils/decoration.py]: Well sounds like extend and append for this hight level case should be handled as a single thing/method/function. I know this was supposed to look like a list but it makes no sense that it has to follow the list primitives, this is a higher level object, not a list.

Comment (rlaverde) [reply spyder/widgets/sourcecode/utils/decoration.py]: Ok, then I'll fusion append and extend into a function named add

File: spyder/widgets/sourcecode/base.py (hunk 3 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/base.py]: I improved other docstrings, but I didn't modify this function at all, fixing that docstring will add noise to the PR

File: spyder/widgets/panels/codefolding.py [hunk 7]
```diff
@@ -348,12 +348,21 @@ def _get_scope_highlight_color(self):
         return color
 
     def _decorate_block(self, start, end):
+        """
+        Create a decoration and addedit to the editor.
```
Comment (ccordoba12) [inline spyder/widgets/panels/codefolding.py]: `addedit` -> `add it`?

File: spyder/widgets/sourcecode/base.py [hunk 8]
```diff
@@ -317,43 +319,61 @@ def set_palette(self, background, foreground):
...
+
+        Args:
+            key (str) name of the extra selections group.
+            extra_selections (list of sourcecode.api.TextDecoration).
+        """
+        # current line has to be highlighted first
+        draw_orders = {'current_cell': 0,
+                       # draw order = 1 is used in codefolding
+                       'current_line': 2}
+        draw_order = draw_orders.get(key,  5)
```
Comment (ccordoba12) [inline spyder/widgets/sourcecode/base.py]: Why `5` here?

File: spyder/widgets/sourcecode/base.py [hunk 9]
```diff
@@ -317,43 +319,61 @@ def set_palette(self, background, foreground):
...
+
+        NOTE: This will remove previous decorations added to  the same key.
+
+        Args:
+            key (str) name of the extra selections group.
+            extra_selections (list of sourcecode.api.TextDecoration).
+        """
+        # current line has to be highlighted first
+        draw_orders = {'current_cell': 0,
+                       # draw order = 1 is used in codefolding
```
Comment (ccordoba12) [inline spyder/widgets/sourcecode/base.py]: Why `0` can't be used for code folding?

File: spyder/widgets/sourcecode/tests/test_extra_selections.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    editor = construct_editor()
+
+    # Move cursor over 'some_variable'
+    editor.go_to_line(2)
+    cursor = editor.textCursor()
+    cursor.movePosition(QTextCursor.Right, n=5)
+    editor.setTextCursor(cursor)
+
+    qtbot.waitUntil(lambda: len(editor.extraSelections()) >= 6, timeout=2000)
+    print(len(editor.extraSelections()))
```
Comment (ccordoba12) [inline spyder/widgets/sourcecode/tests/test_extra_selections.py]: Is this print really needed?

Comment (goanpeca) [reply spyder/widgets/sourcecode/tests/test_extra_selections.py]: I think it ts a good practice to have prints inside tests cause when the test fails you get more debugging info. Dont know of it applies to this though.

File: spyder/widgets/panels/codefolding.py (hunk 7 above)
Comment (rlaverde) [reply spyder/widgets/panels/codefolding.py]: Yes :-)

File: spyder/widgets/sourcecode/base.py (hunk 8 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/base.py]: It just have to be a number greater that current_line and curreent_cell.

I choose 5 to leave 3 and 4 free, maybe somebody will need to add something in front of the current line, but behind if the other selections.

File: spyder/widgets/sourcecode/base.py (hunk 9 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/base.py]: I put 1 to make it appear in front of current cell, but behind other selections.

I'm not sure, what is the best order, codefolding in the back? Or maybe in the front and be a little transparent?

File: spyder/widgets/sourcecode/tests/test_extra_selections.py (hunk 10 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/tests/test_extra_selections.py]: Hmm no, I was using it locally for know the variables when the test failed. I'll remove it

Comment (goanpeca) [reply spyder/widgets/sourcecode/tests/test_extra_selections.py]: >I was using it locally for know the variables when the test failed

Well.. if the test fails, it would be nice to have that information, no? otherwise we just get an assertion error that might not provide enough info

Comment (rlaverde) [reply spyder/widgets/sourcecode/tests/test_extra_selections.py]: This print wasn't giving too much information.

But adding print to test could be a good practice, for easy now what is failing in the CI when a test fails.

File: spyder/widgets/sourcecode/base.py (hunk 8 above)
Comment (ccordoba12) [reply spyder/widgets/sourcecode/base.py]: Please define and use constants for this. Something like:

    CELL_ORDER = 0
    FOLDING_ODER = 1
    LINE_ORDER = 2

and use those constants here.

File: spyder/widgets/sourcecode/base.py (hunk 9 above)
Comment (ccordoba12) [reply spyder/widgets/sourcecode/base.py]: This needs to be explained when you define the constants I mentioned above.

File: spyder/widgets/sourcecode/base.py (hunk 8 above)
Comment (rlaverde) [reply spyder/widgets/sourcecode/base.py]: I'll use this same dict an expose it as a part of the decoration api.

Comment (ccordoba12) [reply spyder/widgets/sourcecode/base.py]: Great!

Comment (rlaverde) [reply spyder/widgets/sourcecode/base.py]: I just add a commit with the DRAW_ORDERS dict, I didn't leave 3 and 4 unused because now its easier to modify the draworders values.

**Comment by ccordoba12** [comment]
This looks really good to me. @goanpeca?
